pkg/witness/publish_checkpoint.go

```diff
@@ -0,0 +1,191 @@
+	redisClient *redis.Client
+	// reqCounter tracks successes and failures for publishing
+	reqCounter *prometheus.CounterVec
+}
+
+// Constant values used with metrics
+const (
+	Success = iota
+	SuccessObtainLock
+	GetCheckpoint
+	UnmarshalCheckpoint
+	SignCheckpoint
+	RedisFailure
+	RedisLatestFailure
+)
+
+// NewCheckpointPublisher creates a CheckpointPublisher to write stable checkpoints to Redis
+func NewCheckpointPublisher(ctx context.Context,
+	logClient trillian.TrillianLogClient,
+	treeID int64,
+	hostname string,
+	signer signature.Signer,
+	redisClient *redis.Client,
+	checkpointFreq uint,
+	reqCounter *prometheus.CounterVec) CheckpointPublisher {
+	return CheckpointPublisher{ctx: ctx, logClient: logClient, treeID: treeID, hostname: hostname,
+		signer: signer, checkpointFreq: checkpointFreq, redisClient: redisClient, reqCounter: reqCounter}
+}
+
+// StartPublisher creates a long-running task that publishes the latest checkpoint every X minutes
+func (c *CheckpointPublisher) StartPublisher() {
+	tc := util.NewTrillianClient(context.Background(), c.logClient, c.treeID)
+	sTreeID := strconv.FormatInt(c.treeID, 10)
+
+	// publish on startup to ensure a checkpoint is available the first time Rekor starts up
+	c.publish(&tc, sTreeID)
+
+	ticker := time.NewTicker(time.Duration(c.checkpointFreq) * time.Minute)
+	go func() {
+		for {
+			<-ticker.C
+			c.publish(&tc, sTreeID)
+		}
+	}()
+}
+
+// publish publishes the latest checkpoint to Redis once
+func (c *CheckpointPublisher) publish(tc *util.TrillianClient, sTreeID string) {
+	// get latest checkpoint
+	resp := tc.GetLatest(0)
+	if resp.Status != codes.OK {
+		c.reqCounter.With(
+			map[string]string{
+				"shard": sTreeID,
+				"code":  strconv.Itoa(GetCheckpoint),
+			}).Inc()
+		log.Logger.Errorf("error getting latest checkpoint to publish: %v", resp.Status)
+		return
+	}
+
+	// unmarshal checkpoint
+	root := &types.LogRootV1{}
+	if err := root.UnmarshalBinary(resp.GetLatestResult.SignedLogRoot.LogRoot); err != nil {
+		c.reqCounter.With(
+			map[string]string{
+				"shard": sTreeID,
+				"code":  strconv.Itoa(UnmarshalCheckpoint),
+			}).Inc()
+		log.Logger.Errorf("error unmarshalling latest checkpoint to publish: %v", err)
+		return
+	}
+
+	// sign checkpoint with Rekor private key
+	checkpoint, err := util.CreateAndSignCheckpoint(context.Background(), c.hostname, c.treeID, root, c.signer)
+	if err != nil {
+		c.reqCounter.With(
+			map[string]string{
+				"shard": sTreeID,
+				"code":  strconv.Itoa(SignCheckpoint),
+			}).Inc()
+		log.Logger.Errorf("error signing checkpoint to publish: %v", err)
+		return
+	}
+
+	// encode checkpoint as hex to write to redis
+	hexCP := hex.EncodeToString(checkpoint)
+
+	// write checkpoint to Redis if key does not yet exist
+	// this prevents multiple instances of Rekor from writing different checkpoints in the same time window
+	ts := time.Now().Truncate(time.Duration(c.checkpointFreq) * time.Minute).UnixNano()
```

Comment:
Update to only use as locking mechanism, don't need to store checkpoint too

Reply:
Decided to leave this as is, as it gives us a history of checkpoints if need be.